<a href="https://colab.research.google.com/github/kabirulislam95/AI/blob/main/Assainment04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/kabirulislam95/AI/blob/main/Assainment04.ipynb)

In [ ]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.datasets import load_iris
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier


### **Part1: Load the dataset**

In [ ]:
# Importing the dataset
data = sns.load_dataset("iris")
#print top 15 of the dataset
data.head(15)

sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8          4.0           1.2          0.2  setosa

In [ ]:
#print the summary of the dataset
data.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

### **About the dataset**
for each data it has four different variables, sepal_length, sepal_width, petal_length and petal_width. Based of these four variables, each flower belongs to a particular species. In our dataset, it has three dirrent types of species, setosa, versicolor and virginica.

The four variables are the features and 'species' is the labels. Based on the four features, each new data would be label as setaso or versicolor or virginica.

### **Part2: Split the dataset**

In [ ]:
# Splitting the dataset into the Training set and Test set
# X = feature values, all the columns except the last column
X = data.iloc[:, :-1]

# y = target/label values, last column of the dataset
y = data.iloc[:, -1]

In [ ]:
#print first 5 features value
X.head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [ ]:
#print first 5 label values
y.head()

0    setosa
1    setosa
2    setosa
3    setosa
4    setosa
Name: species, dtype: object

In [ ]:
#Split the data into 90% training and 10% testing
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)

### **Part3: Logistic Regression**

In [ ]:
#take logostic regression model object
model_lregression = LogisticRegression(max_iter=1000)
#Training the model
model_lregression.fit(x_train.values, y_train.values) 

LogisticRegression(max_iter=1000)

In [ ]:
#probability of each species
pd.DataFrame(model_lregression.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.052595    0.017443   0.929961

In [ ]:
#Creating samples
sample1 = np.array([[4.9, 3.0, 1.5, 0.2]])
sample2 = np.array([[4.9, 2.5, 3.2, 1.1]])
sample3 = np.array([[4.9, 2.9, 4.5, 1.9]])

In [ ]:
#predict sample1
model_lregression.predict(sample1)

array(['setosa'], dtype=object)

In [ ]:
#predict sample2
model_lregression.predict(sample2)

array(['versicolor'], dtype=object)

In [ ]:
#predict sample3
model_lregression.predict(sample3)

array(['virginica'], dtype=object)

In [ ]:
#find accuracy score
model_lregression.score(x_test.values, y_test)

1.0

The score is 1 there is a 1*100 = 100% chance of the classification being correct

In [ ]:
#coefficient
model_lregression.coef_

array([[-0.41087786,  0.92630746, -2.43464502, -1.04770092],
       [ 0.40315462, -0.32623036, -0.11363406, -0.88710526],
       [ 0.00772324, -0.6000771 ,  2.54827907,  1.93480618]])

In [ ]:
#intercept
model_lregression.intercept_

array([  9.46338452,   2.44534908, -11.9087336 ])

### **Part4: Support Vector Machine**

In [ ]:
#create SVM model object
model_svm=svm.SVC(kernel = 'linear', probability=True, C=1)

In [ ]:
#train the model
model_svm.fit(x_train.values, y_train.values)

SVC(C=1, kernel='linear', probability=True)

In [ ]:
#probability of each species
pd.DataFrame(model_svm.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.045702    0.004418    0.94988

In [ ]:
#create sample dataset
sample1 = np.array([[4.9, 3.0, 1.5, 0.2]])
sample2 = np.array([[4.9, 2.5, 3.2, 1.1]])
sample3 = np.array([[4.9, 2.9, 4.5, 1.9]])

In [ ]:
#predict sample1
model_svm.predict(sample1)

array(['setosa'], dtype=object)

In [ ]:
#predict sample2
model_svm.predict(sample2)

array(['versicolor'], dtype=object)

In [ ]:
#predict sample3
model_svm.predict(sample3)

array(['virginica'], dtype=object)

In [ ]:
#score of SVM model
model_svm.score(x_test.values, y_test)

1.0

The score is 1 there is a 1*100 = 100% chance of the classification being correct

### **Part5: Neural Network**

In [ ]:
# take MLP classifier Model and train the model with training dataset
model_neural_lbfgs = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(x_train.values, y_train.values)

In [ ]:
# print the probability of the model 
pd.DataFrame(model_neural_lbfgs.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor     Virginica
0     0.0         1.0  6.632171e-44

In [ ]:
#creating sample
sample1 = np.array([[4.9, 3.0, 1.5, 0.2]])
sample2 = np.array([[4.9, 2.5, 3.2, 1.1]])
sample3 = np.array([[4.9, 2.9, 4.5, 1.9]])

In [ ]:
#predict sample1
model_neural_lbfgs.predict(sample1)

array(['setosa'], dtype='<U10')

In [ ]:
#predict sample2
model_neural_lbfgs.predict(sample2)

array(['versicolor'], dtype='<U10')

In [ ]:
#predict sample3
model_neural_lbfgs.predict(sample3)

array(['virginica'], dtype='<U10')

In [ ]:
#print score of the model
model_neural_lbfgs.score(x_test.values,y_test)

1.0

The score is 1 there is a 1*100 = 100% chance of the classification being correct

In [ ]:
# Experiment with different option for the Neural Network
# for SGD
model_neural_sgd = MLPClassifier(hidden_layer_sizes=(50,), max_iter=10, alpha=1e-4,solver='sgd', verbose=10, tol=1e-4, random_state=1,learning_rate_init=.1)
model_neural_sgd.fit(x_train.values, y_train.values)

Iteration 1, loss = 1.30753661
Iteration 2, loss = 2.24888506
Iteration 3, loss = 1.07653202
Iteration 4, loss = 1.32868279
Iteration 5, loss = 1.25519597
Iteration 6, loss = 0.78373787
Iteration 7, loss = 0.58890651
Iteration 8, loss = 0.46293972
Iteration 9, loss = 0.41887034
Iteration 10, loss = 0.40162216


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


MLPClassifier(hidden_layer_sizes=(50,), learning_rate_init=0.1, max_iter=10,
              random_state=1, solver='sgd', verbose=10)

In [ ]:
# print probability for SGD
pd.DataFrame(model_neural_sgd.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.004611    0.264172   0.731216

In [ ]:
# print score for SGD
model_neural_sgd.score(x_test.values, y_test)

0.6

In [ ]:
# For Adam
model_neural_adam = MLPClassifier(hidden_layer_sizes=(50,), max_iter=10, alpha=1e-4,solver='adam', verbose=10, tol=1e-4, random_state=1,learning_rate_init=.1)
model_neural_adam.fit(x_train.values, y_train.values)

Iteration 1, loss = 1.30753661
Iteration 2, loss = 3.06026818
Iteration 3, loss = 1.77629091
Iteration 4, loss = 0.69693028
Iteration 5, loss = 1.37180009
Iteration 6, loss = 0.75713206
Iteration 7, loss = 0.61692559
Iteration 8, loss = 0.48235648
Iteration 9, loss = 0.47512579
Iteration 10, loss = 0.48018509


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


MLPClassifier(hidden_layer_sizes=(50,), learning_rate_init=0.1, max_iter=10,
              random_state=1, verbose=10)

In [ ]:
# Probability for adam
pd.DataFrame(model_neural_adam.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.000022    0.141723   0.858255

In [ ]:
# print score for Adam
model_neural_adam.score(x_test.values, y_test)

0.9333333333333333

We see that  for lbfgs we got 100% accuracy, for SGD 60% and for adam 93%. So, we will choose the lbfgs in this case.

### **Part6: K-Nearest Neighbors**

In [ ]:
# create KNeighborClassifier model object
model_knn = KNeighborsClassifier(n_neighbors = 3)

#train the model
model_knn.fit(x_train.values, y_train.values)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
#probability of each species
pd.DataFrame(model_knn.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0     0.0    0.666667   0.333333

In [ ]:
#create sample dataset
sample1 = np.array([[4.9, 3.0, 1.5, 0.2]])
sample2 = np.array([[4.9, 2.5, 3.2, 1.1]])
sample3 = np.array([[4.9, 2.9, 4.5, 1.9]])

In [ ]:
#predict sample1
model_knn.predict(sample1)

array(['setosa'], dtype=object)

In [ ]:
#predict sample2
model_knn.predict(sample2)

array(['versicolor'], dtype=object)

In [ ]:
#predict sample3
model_knn.predict(sample3)

array(['versicolor'], dtype=object)

In [ ]:
# print score of the model
model_knn.score(x_test.values, y_test)

1.0

### **Part7: Conclusion and Takeaways**

We see that we got different probabilities in different model, but almost all the models we got 100% accuracy. So either one we can choose. The probability and the scores are given below:

**Logistic regression:**

**probability:**

           Setosa = 0.052595

            Versicolor = 0.017443

             Virginica = 0.929961

  Model Score = 100%



  **Support Vector Machine:**

**probability:**

           Setosa = 0.045702

            Versicolor = 0.004418

             Virginica = 0.94988

  Model Score = 100%



**Neural Network:**

**probability for lbfgs:**

           Setosa = 0.0

           Versicolor = 1.0

Virginica = 6.632171$e^{-44}$ 
             
             
  Model Score = 100%


**probability for SGD:**

           Setosa = 0.004611

            Versicolor = 0.264172

             Virginica = 0.731216

  Model Score = 60%


  **probability for Adam:**

           Setosa = 0.000022

            Versicolor = 0.141723

             Virginica = 0.858255

  Model Score = 93%


**K Nearest Neighbors:**

**probability:**

           Setosa = 0.0

            Versicolor = 0.666667

             Virginica = 0.333333

  Model Score = 100%


So, we can choose any model. But for easy and most popular implementation is the logistic regression. So, I will choose this model. On the other hand if we shuffle the dataset the value could be different. In our program we don't fix the random state, we can get different result each run.
